### Cameron Stewart
# Project 3

### 1.	Compare your given name with your nickname (if you don’t have a nickname, invent one for this assignment) by answering the following questions:
*   What is the edit distance between your nickname and your given name?
*	What is the percentage string match between your nickname and your given name?
*   Show your work for both calculations.

First, I found the edit distance between my name and nickname

In [1]:
#Import required libraries
import nltk

In [2]:
#Names to compare
given_name='Cameron'
nickname='Camo'

#Levenshtein edit distance between given and nickname
edit_distance=nltk.edit_distance(given_name,nickname)
print('Given Name:',given_name)
print('Nickname',nickname)
print('Edit Distance:',edit_distance)
print("Edit Steps: 'Cameron'->'Camero'->'Cameo'->'Camo'")

Given Name: Cameron
Nickname Camo
Edit Distance: 3
Edit Steps: 'Cameron'->'Camero'->'Cameo'->'Camo'


Next, I found the percentage string match between my name and nickname

In [3]:
#Import required libraries
import textdistance

In [4]:
#Names to compare
given_name='Cameron'
nickname='Camo'

#String match between given and nickname
edit_distance=nltk.edit_distance(given_name,nickname)
string_match_perc=textdistance.levenshtein.normalized_similarity(given_name,nickname)
print('Given Name:',given_name)
print('Nickname:',nickname)
print("Edit Steps: 'Cameron'->'Camero'->'Cameo'->'Camo'")
print("String Match Percentage = 1-(edit distance/length of longer string)")
print("String Match Percentage = 1-(",edit_distance,"/",max(len(given_name),len(nickname)),")",sep="")
print('String Match Percentage: ',round(string_match_perc*100,2),"%",sep="")

Given Name: Cameron
Nickname: Camo
Edit Steps: 'Cameron'->'Camero'->'Cameo'->'Camo'
String Match Percentage = 1-(edit distance/length of longer string)
String Match Percentage = 1-(3/7)
String Match Percentage: 57.14%


### 2.	Find a friend (or family member or classmate) who you know has read a certain book. Without your friend knowing, copy the first two sentences of that book. Now rewrite the words from those sentences, excluding stop words. Now tell your friend to guess which book the words are from by reading them just that list of words. Did you friend correctly guess the book on the first try? What did he or she guess? Explain why you think you friend either was or was not able to guess the book from hearing the list of words. 

In [5]:
from nltk.corpus import stopwords

In [6]:
#Storing and tokenizing opening text
opening_text="When I think of my wife, I always think of her head. The shape of it, to begin with."
opening_text=opening_text.lower()
tokenized_text=nltk.word_tokenize(opening_text)

#Remove stop words and compare
tokenized_text_stop_removed=[x for x in tokenized_text if x not in stopwords.words('english')]
print('Original Text:',tokenized_text)
print('Opening Text With Stop Words Removed:', tokenized_text_stop_removed)

Original Text: ['when', 'i', 'think', 'of', 'my', 'wife', ',', 'i', 'always', 'think', 'of', 'her', 'head', '.', 'the', 'shape', 'of', 'it', ',', 'to', 'begin', 'with', '.']
Opening Text With Stop Words Removed: ['think', 'wife', ',', 'always', 'think', 'head', '.', 'shape', ',', 'begin', '.']


The book selected to parse out the stop words was 'Gone Girl' by Gillian Flynn. The book guessed was Every Breath You Take: A True Story of Obsession, Revenge, and Murder by Ann Rule. After a few more guesses and light hinting, she was able to guess the correct book. I don't believe she was able to guess the correct book because she reads many murder mystery books and the first few sentences are not that specific. The interesting takeaway was how much insight she was able to obtain from such a small amount of information. She had a sense it was a murder mystery genre novel about a husband and wife from just 7 words. I imagine with a few more sentences that she would have been able to correctly guess on the first attempt.

### 3.	Run one of the stemmers available in Python. Run the same two sentences from question 2 above through the stemmer and show the results. How many of the outputted stems are valid morphological roots of the corresponding words? Express this answer as a percentage.

In [7]:
import string

In [8]:
#Storing and tokenizing opening text
opening_text="When I think of my wife, I always think of her head. The shape of it, to begin with."
opening_text=opening_text.lower()
tokenized_text=nltk.word_tokenize(opening_text)

#Remove stop words
tokenized_text_stop_removed=[x for x in tokenized_text if x not in stopwords.words('english')]

#Stem the non-punctuation tokens
stemmed_words_w_sw=[nltk.PorterStemmer().stem(word) for word in tokenized_text if word not in string.punctuation]
stemmed_words_wo_sw=[nltk.PorterStemmer().stem(word) for word in tokenized_text_stop_removed if word not in string.punctuation]
print('Stemmed Opening Text With Porter Stemmer:',stemmed_words_w_sw)

#We will find the morphological root using lemmatization on the non-punctuation tokens
lemmatized_words_w_sw=[nltk.WordNetLemmatizer().lemmatize(word) for word in tokenized_text if word not in string.punctuation]
lemmatized_words_wo_sw=[nltk.WordNetLemmatizer().lemmatize(word) for word in tokenized_text_stop_removed if word not in string.punctuation]
print('Lemmatized Opening Text With WordNet Lemmatizer:',lemmatized_words_w_sw)

#Find the percentage of stems that are valid morphological roots
morph_root_percentage_w_sw=sum([lemmatized_words_w_sw[x]==stemmed_words_w_sw[x] for x in range(len(lemmatized_words_w_sw))])/len(lemmatized_words_w_sw)
morph_root_percentage_wo_sw=sum([lemmatized_words_wo_sw[x]==stemmed_words_wo_sw[x] for x in range(len(lemmatized_words_wo_sw))])/len(lemmatized_words_wo_sw)
print('Valid Morphological Root Percentage WITH Stop Words: ',round(morph_root_percentage_w_sw*100,2),'%',sep="")
print('Valid Morphological Root Percentage WITHOUT Stop Words: ',round(morph_root_percentage_wo_sw*100,2),'%',sep="")

Stemmed Opening Text With Porter Stemmer: ['when', 'i', 'think', 'of', 'my', 'wife', 'i', 'alway', 'think', 'of', 'her', 'head', 'the', 'shape', 'of', 'it', 'to', 'begin', 'with']
Lemmatized Opening Text With WordNet Lemmatizer: ['when', 'i', 'think', 'of', 'my', 'wife', 'i', 'always', 'think', 'of', 'her', 'head', 'the', 'shape', 'of', 'it', 'to', 'begin', 'with']
Valid Morphological Root Percentage WITH Stop Words: 94.74%
Valid Morphological Root Percentage WITHOUT Stop Words: 85.71%


I made the assumption to remove punctuation as tokens in this analysis of stemming and valid morphological root words. I used the Porter Stemmer to find the stem of each word. I used the WordNet Lemmatizer to find the valid mophological root of each word. After comparing these two transformations of the text, the only word that is different is the 'alway' stem and 'always' lemma. This one difference resulted in a 94.74% success rate of the stemmer finding a valid morphological root. This value significantly changes to 85.71% if you remove the stop words.